### PR TITLE
Bitwise ops

### PR DESCRIPTION
**Pull request by hameerabbasi**
Added bitwise operators and made changes so comparisons with zero are robust.

### CODE REVIEW

**Comment by mrocklin** [comment]
It looks like this has your previous commits in it and needs to be rebased off of master.  We tend to use the squash-and-merge PR policy.

**Comment by hameerabbasi** [comment]
I'm not very good ad rebasing. Apologies for the extra commits.

**Comment by hameerabbasi** [comment]
If it isn't too much to ask, I'd like a link to a resource on how to avoid things like this in the future.

**Comment by mrocklin** [comment]
Here are a couple links:

- https://guides.github.com/introduction/flow/
- https://pandas.pydata.org/pandas-docs/stable/contributing.html#working-with-the-code

I think that the lesson to learn in this situation is to never develop in your master branch (like you did your last PR) but instead always work in a development branch (like you are here) so that you always have a fresh version to pull down recent changes.  I usually do the following:

```
git checkout master
git pull upstream master
git checkout -b my-new-branch
...
git commit -a
git push mrocklin my-new-branch
```

To resolve your particular issue I would normally do the following:

```
# you may already have these, but people use different names so I'm adding these to be explicit
git remote add mrocklin git@github.com:mrocklin/sparse.git
git remote add hameerabbasi git@github.com:hameerabbasi/sparse.git

# change your master branch to exactly reflect the current master in the "main" fork (in this case, mine, but usually something like pydata or dask)
git checkout master
git remote update mrocklin
git reset --hard mrocklin/master  # WARNING: this will destroy any information you may have on your current master branch

# Rebase your development branch on to the master branch (this will require understanding a bit about rebase
git checkout bitwise_ops
git rebase -i master # clean up history here a bit, possibly removing all commits that aren't part of this branch

git push -f hameerabbasi bitwise_ops
```

If you don't understand rebase then I can also clean things up and force-push to your fork

**Comment by hameerabbasi** [comment]
Okay, cleaned up the commit history. Should be good to go.

File: sparse/core.py [hunk 1]
```diff
@@ -701,12 +719,16 @@ def elemwise_binary(self, func, other, *args, **kwargs):
...
-        nonzero = data != 0
+        nonzero = data != func_zero
         data = data[nonzero]
         coords = coords[:, nonzero]
 
         return COO(coords, data, shape=result_shape, has_duplicates=False)
 
+    @staticmethod
+    def _zero_of_dtype(dtype):
+        return np.zeros((), dtype=dtype)
```
Comment (mrocklin) [inline sparse/core.py]: I have a slight preference to move utility functions like this to be normal functions rather than staticmethods.

Comment (hameerabbasi) [reply sparse/core.py]: Would you prefer it be in `utils.py` or `core.py`?

**Comment by mrocklin** [comment]
Either.  No preference there

**Comment by mrocklin** [comment]
I would probably put it in core just for locality

**Comment by hameerabbasi** [comment]
Done. If you haven't spotted anything else, this should be good to merge.

**Comment by mrocklin** [comment]
This is in.  Thanks @hameerabbasi !

**Pull request merged**